---
title: "PCOA"
output: html_document
date: "2024-11-14"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/alex/Documents/Documents - Alex’s MacBook Pro/Documents/GitHub/Aus-Invasions-2023-Course")
setwd("/Users/alex/Documents/Documents - Alex’s MacBook Pro/Documents/GitHub/Aus-Invasions-2023-Course")
```

#Load packages
```{r}
library("tidyverse")
library("mctoolsr")
library("knitr")
library(ggplot2)
library(vegan)
```

#Starting with 16S
#Loading in files
```{r}
zotutab1 <- read_delim("Aus23_16S_Metabarcoding/Aus23_16S_ASV_table.txt",delim = "\t")
meta1 <- read_delim("Aus23_16S_Metabarcoding/MetabarcodingMetadata.txt", delim = "\t")
tax1 <- read_delim("Aus23_16S_Metabarcoding/Aus23_16S_GGTaxonomy_16S.txt", delim = "\t", col_names = FALSE)
tax1 <- tax1[-1,-3]
taxcut1 <- tax1 %>%
  dplyr::rename("#OTU ID" = "X1", "taxonomy" = "X2")
```

#Removing chloroplast/mitochondria reads
```{r}
#Calculating OTU Totals
otutotals1 <- mutate(.data = zotutab1, otutotal = rowSums(zotutab1[,2:74]))

#Merge taxonomy info
otutabtax1 <- inner_join(otutotals1, taxcut1, by = "#OTU ID")

otutabtax1$host <- 0

otuhost1 <- otutabtax1 %>%
  mutate(host = case_when(
    grepl("Rickettsiales", taxonomy) ~ "remove",
    grepl("Chloroplast", taxonomy) ~ "remove",
    TRUE ~ "keep"))

otufilt1 <- filter(.data = otuhost1, host == "keep")

mcotu1 <- otufilt1 %>%
  dplyr::select(-otutotal, -host)
```

#Writing delim files for mctools
```{r}
write_delim(mcotu1, file = "mcotu1.txt", delim = "\t")
write_delim(meta1, file = "meta1.txt", delim = "\t")

#Loading into mctoolsr
input <- load_taxa_table("mcotu1.txt", "meta1.txt")
```

#Filtering to root or soil only
```{r}
root = filter_data(input, filter_cat = "SampleType", keep_vals = "Root")
soil = filter_data(input, filter_cat = "SampleType", keep_vals = "Soil")
```

#Root PCoA
```{r}
#Calculate PCoA
dm <- calc_dm(root$data_loaded)
#PCoA
ord1 <- calc_ordination(dm, 'pcoa')
plot_ordination(root, ord1, 'TreeSpecies', hulls = TRUE)
```

#Soil PCoA
```{r}
#Calculate PCoA
dm <- calc_dm(soil$data_loaded)
#PCoA
ord1 <- calc_ordination(dm, 'pcoa')
plot_ordination(root, ord1, 'TreeSpecies', hulls = TRUE)
```

#Moving to ITS
```{r}
zotutab1 <- read_delim("/Users/alex/Documents/Documents - Alex’s MacBook Pro/Documents/My Projects/Australia/ITS2_OTUTable_97.txt", delim = "\t")
meta1 <- read_delim("Aus23_16S_Metabarcoding/MetabarcodingMetadata.txt", delim = "\t")
tax1 <- read_delim("/Users/alex/Documents/Documents - Alex’s MacBook Pro/Documents/My Projects/Australia/ITS2_Dada2_repseqs97_taxonomy.txt", delim = "\t", col_names = FALSE)
tax1 <- tax1[-1,-3]
taxcut1 <- tax1 %>%
  dplyr::rename("#OTU ID" = "X1", "taxonomy" = "X2")
```

#Adjusting dfs (not necessary but keeps naming conventions)
```{r}
#Calculating OTU Totals
otutotals1 <- mutate(.data = zotutab1, otutotal = rowSums(zotutab1[,2:74]))

#Merge taxonomy info
otutabtax1 <- inner_join(otutotals1, taxcut1, by = "#OTU ID")

otutabtax1$host <- 0

otuhost1 <- otutabtax1 %>%
  mutate(host = case_when(
    grepl("Rickettsiales", taxonomy) ~ "remove",
    grepl("Chloroplast", taxonomy) ~ "remove",
    TRUE ~ "keep"))

otufilt1 <- filter(.data = otuhost1, host == "keep")

mcotu1 <- otufilt1 %>%
  dplyr::select(-otutotal, -host)
```

#Writing delim files for mctoolsr
```{r}
write_delim(mcotu1, file = "mcotu1.txt", delim = "\t")
write_delim(meta1, file = "meta1.txt", delim = "\t")

#Loading into mctoolsr
input <- load_taxa_table("mcotu1.txt", "meta1.txt")
```

#Filtering to root or soil only
```{r}
root = filter_data(input, filter_cat = "SampleType", keep_vals = "Root")
soil = filter_data(input, filter_cat = "SampleType", keep_vals = "Soil")
```

#Root PCoA
```{r}
#Calculate PCoA
dm <- calc_dm(root$data_loaded)
#PCoA
ord1 <- calc_ordination(dm, 'pcoa')
plot_ordination(root, ord1, 'TreeSpecies', hulls = TRUE)
```

#Soil PCoA
```{r}
#Calculate PCoA
dm <- calc_dm(soil$data_loaded)
#PCoA
ord1 <- calc_ordination(dm, 'pcoa')
plot_ordination(root, ord1, 'TreeSpecies', hulls = TRUE)
```

#Moving to 18S
```{r}
zotutab1 <- read_delim("/Users/alex/Documents/Documents - Alex’s MacBook Pro/Documents/My Projects/Australia/AMF_FeatureTable.txt", delim = "\t")
meta1 <- read_delim("Aus23_16S_Metabarcoding/MetabarcodingMetadata.txt", delim = "\t")
tax1 <- read_delim("/Users/alex/Documents/Documents - Alex’s MacBook Pro/Documents/My Projects/Australia/AMF_Taxonomy.txt", delim = "\t", col_names = FALSE)
tax1 <- tax1[-1,-3]
taxcut1 <- tax1 %>%
  dplyr::rename("#OTU ID" = "X1", "taxonomy" = "X2")
```

```{r}
zotutab1 <- read_delim("/Users/alex/Documents/Documents - Alex’s MacBook Pro/Documents/My Projects/Australia/ITS2_OTUTable_97.txt", delim = "\t")
meta1 <- read_delim("Aus23_16S_Metabarcoding/MetabarcodingMetadata.txt", delim = "\t")
tax1 <- read_delim("/Users/alex/Documents/Documents - Alex’s MacBook Pro/Documents/My Projects/Australia/ITS2_Dada2_repseqs97_taxonomy.txt", delim = "\t", col_names = FALSE)
tax1 <- tax1[-1,-3]
taxcut1 <- tax1 %>%
  dplyr::rename("#OTU ID" = "X1", "taxonomy" = "X2")
```

#Adjusting dfs (not necessary but keeps naming conventions)
```{r}
#Calculating OTU Totals
otutotals1 <- mutate(.data = zotutab1, otutotal = rowSums(zotutab1[,2:73]))

#Merge taxonomy info
otutabtax1 <- inner_join(otutotals1, taxcut1, by = "#OTU ID")

otutabtax1$host <- 0

otuhost1 <- otutabtax1 %>%
  mutate(host = case_when(
    grepl("Rickettsiales", taxonomy) ~ "remove",
    grepl("Chloroplast", taxonomy) ~ "remove",
    TRUE ~ "keep"))

otufilt1 <- filter(.data = otuhost1, host == "keep")

mcotu1 <- otufilt1 %>%
  dplyr::select(-otutotal, -host)
```

#Writing delim files for mctoolsr
```{r}
write_delim(mcotu1, file = "mcotu1.txt", delim = "\t")
write_delim(meta1, file = "meta1.txt", delim = "\t")

#Loading into mctoolsr
input <- load_taxa_table("mcotu1.txt", "meta1.txt")
```

#Filtering to root or soil only
```{r}
root = filter_data(input, filter_cat = "SampleType", keep_vals = "Root")
soil = filter_data(input, filter_cat = "SampleType", keep_vals = "Soil")
```

#Root PCoA
```{r}
#Calculate PCoA
dm <- calc_dm(root$data_loaded)
#PCoA
ord1 <- calc_ordination(dm, 'pcoa')
plot_ordination(root, ord1, 'TreeSpecies', hulls = TRUE)
```

#Soil PCoA
```{r}
#Calculate PCoA
dm <- calc_dm(soil$data_loaded)
#PCoA
ord1 <- calc_ordination(dm, 'pcoa')
plot_ordination(root, ord1, 'TreeSpecies', hulls = TRUE)
```

